replication-api-impl/src/main/java/com/connexta/replication/api/impl/data/FilterIndexImpl.java

```diff
@@ -0,0 +1,88 @@
+/**
+ * Copyright (c) Connexta
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package com.connexta.replication.api.impl.data;
+
+import com.connexta.ion.replication.api.ReplicationPersistenceException;
+import com.connexta.replication.api.data.Filter;
+import com.connexta.replication.api.data.FilterIndex;
+import com.connexta.replication.api.impl.persistence.pojo.FilterIndexPojo;
+import java.time.Instant;
+import java.util.Optional;
+
+/** Simple implementation of FilterIndex. */
+public class FilterIndexImpl extends AbstractPersistable<FilterIndexPojo> implements FilterIndex {
+
+  private static final String TYPE = "filter_index";
+
+  private Instant modifiedSince;
+
+  /**
+   * Creates a new index for a Filter.
+   *
+   * @param filter to create an index for
+   */
+  public FilterIndexImpl(Filter filter) {
+    super(FilterIndexImpl.TYPE, filter.getId());
+  }
+
+  /**
+   * Creates a new FilterIndex.
+   *
+   * @param pojo to create the filter index from
+   */
+  protected FilterIndexImpl(FilterIndexPojo pojo) {
+    super(FilterIndexImpl.TYPE, null);
+    readFrom(pojo);
+  }
+
+  @Override
+  public Optional<Instant> getModifiedSince() {
+    return Optional.ofNullable(modifiedSince);
+  }
+
+  @Override
+  public void setModifiedSince(Instant modifiedSince) {
```

Comment:
❓ Since we are saying that the modified since is optional, shouldn't we have a setter that accepts `Optional<Instant>` as well?

Reply:
No, if they pass in an Instant or null, the getter will handle the optional part but I should put nullable on the argument here.

Reply:
We should avoid using @nullable on API methods. It is preferable to overload it with an Optional flavor and leave the one there to not accept null.

Reply:
If I do that then I should check for null being passed in here and throw a runtime exception if so. Why not just let the caller pass in null, say for instance to reset the filter's modified since?

Reply:
no need to check for null since the prototype doesn't annotate the parameter with `@Nullable` which by convention means you should be passing it in.

It is a matter of keeping the API clean and avoid using `null` as a valid value which tends to create problems and complexity in client's code. This is one of the main reason why the `Optional` class was introduced

Reply:
Do we not want a way for the client to reset the modified time from the API?

Reply:
`setModifiedSince(Optional.empty())`